### PR TITLE
Implement both Bulk-close endpoints

### PR DESCRIPTION
**Pull request by LBHCallumM**

### CODE REVIEW

File: RepairsApi.Tests/V2/Controllers/BackOfficeControllerTests.cs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            _bulkCloseWorkOrdersUseCaseMock = new Mock<IBulkCloseWorkOrdersUseCase>();
+            _bulkCloseWorkOrdersCloseToBaseUseCaseMock = new Mock<IBulkCloseWorkOrdersCloseToBaseUseCase>();
+
+            _classUnderTest = new BackOfficeController(
+                _bulkCloseWorkOrdersUseCaseMock.Object,
+                _bulkCloseWorkOrdersCloseToBaseUseCaseMock.Object);
+        }
+
+        [Test]
+        public async Task BulkCancelWorkOrders_WhenNoWorkOrderReferencesInRequest_ReturnsBadRequest()
```
Comment (Asura5) [inline RepairsApi.Tests/V2/Controllers/BackOfficeControllerTests.cs:33]: Should this be "when there's no reason"?

File: RepairsApi/V2/Controllers/BackOfficeController.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+        [HttpPost]
+        [Route("bulk-close/cancel")]
+        [ProducesResponseType(typeof(OperativeResponse), 204)]
+        [ProducesResponseType(typeof(string), StatusCodes.Status400BadRequest)]
+        [ProducesDefaultResponseType]
+        // [Authorize(Roles = UserGroups.Agent + "," + UserGroups.ContractManager + "," + UserGroups.AuthorisationManager)]
+        public async Task<IActionResult> BulkCancelWorkOrders([FromBody] BulkCloseWorkOrdersRequestObject request)
+        {
+            if (!request.WorkOrderReferences.Any())
```
Comment (Asura5) [inline RepairsApi/V2/Controllers/BackOfficeController.cs:39]: How about some explicit reason validation?

File: RepairsApi/V2/Gateways/BackOfficeGateway.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+            try
+            {
+                await _repairsContext.Database.ExecuteSqlRawAsync("CALL cancel_old_work_orders(@WorkOrderReferences, @Reason);", workOrderReferences, reason);
+
+                _logger.LogInformation("Successfully called BulkCloseWorkOrders for {Reason}", request.Reason);
+            }
+            catch (Exception e)
+            {
+                _logger.LogInformation("Failed to call BulkCloseWorkOrders - recieving {Error} for {Reason}", e.Message, request.Reason);
```
Comment (Asura5) [inline RepairsApi/V2/Gateways/BackOfficeGateway.cs:40]: You can include the stack trace in CloudWatch by just including e

File: RepairsApi/V2/Helpers/SanitizationHelper.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+{
+    public static class SanitizationHelper
+    {
+        public static string CleanString(string dirtyString)
+        {
+            var invalidCharacters = new HashSet<char>("?&^$#@!()+-,:;<>’\'\"-_*{}[]");
+
+            var result = new StringBuilder(dirtyString.Length);
+
+            foreach (char c in dirtyString)
```
Comment (Asura5) [inline RepairsApi/V2/Helpers/SanitizationHelper.cs:14]: More efficient way (fewer loops):

```
var charsToRemove = new string[] { "@", ",", ".", ";", "'" };
foreach (var c in charsToRemove)
{
    str = str.Replace(c, string.Empty);
}
```

File: RepairsApi.Tests/V2/Controllers/BackOfficeControllerTests.cs (hunk 1 above)
Comment (LBHCallumM) [reply RepairsApi.Tests/V2/Controllers/BackOfficeControllerTests.cs:36]: I did use the `[Required]` attribute, but I have added more explicit validation.

File: RepairsApi/V2/Helpers/SanitizationHelper.cs (hunk 4 above)
Comment (LBHCallumM) [reply RepairsApi/V2/Helpers/SanitizationHelper.cs]: Good idea

**Review by Asura5 — APPROVED** [review]

**Pull request merged**